src/graphql.tsx

```diff
@@ -276,14 +300,27 @@ export default function graphql(
         return opts;
       };
 
-      calculateResultProps(result) {
-        let name = this.type === DocumentType.Mutation ? 'mutate' : 'data';
-        if (operationOptions.name) name = operationOptions.name;
-
-        const newResult = { [name]: result, ownProps: this.props };
-        if (mapResultToProps) return mapResultToProps(newResult);
+      calculateResultProps<T>(result: T) {
+        // Ugly, but the hope is to allow typescript to do control-flow analysis
+        // to determine if `data` or `mutate` are the keys
+        if (operationOptions.name != null) {
+          let name = operationOptions.name;
+          const newResult = { [name]: result, ownProps: this.props };
+          // Prevents us inferring useful type information :/
+          if (mapResultToProps) return mapResultToProps<typeof newResult, { [i: string]: any }>(newResult);
```

Comment:
This is currently preventing us from inferring the type of the result props since the user can feasibly map them to whatever they desire. Perhaps there is a niftier way to do this?